File: 0131_Palindrome_Partitioning.java
class Solution {
    private boolean isPalindrome(String s, int i, int j) {
        while(i < j) {
            if(s.charAt(i) != s.charAt(j)) return false;
            i++; j--;
        }
        return true;
    }
    
    public List<List<String>> partition(String s) {
        Object[] resultAtIndex = new Object[s.length() + 1]; // Obj => List<List<String>>
        
        resultAtIndex[s.length()] = new ArrayList<List<String>>();
        ((List<List<String>>) resultAtIndex[s.length()]).add(new ArrayList<>());
        
        for(int i = s.length() - 1; i >= 0; i--) {
            List<List<String>> currentResult = new ArrayList<>();
            
            for(int j = i; j < s.length(); j++) {
                if(!isPalindrome(s, i, j)) continue;
                
                for(List<String> partitionEnd : (List<List<String>>) resultAtIndex[j + 1]) {
                    ArrayList<String> partition = new ArrayList<String>();
                    partition.add(s.substring(i, j + 1));
                    partition.addAll(partitionEnd);
                    currentResult.add(partition);
                }
            }
            resultAtIndex[i] = currentResult;
        }
        
        return (List<List<String>>) resultAtIndex[0];
    }
}

File: 0131_Palindrome_Partitioning.cpp
class Solution {
public:
    vector<vector<string>> result;
    vector<string> current_partition;
    
    vector<vector<string>> partition(string s) {
        const int n = s.length();
        vector<vector<bool>> isPalindrome(n, vector<bool>(n));
        
        // fill isPalindrome
        for(int i = 0; i < n; i++) for(int j = 0; j < n; j++) {
            int l = i, r = j;
            while(l < r && s[l] == s[r]) l++, r--;
            isPalindrome[i][j] = l >= r;
        }
        
        solve(s, 0, isPalindrome);
        return result;
    }
    
    void solve(string&s, int i, vector<vector<bool>>& isPalindrome) {
        if(i == s.length()) result.push_back(current_partition);
        for(int j = i; j < s.length(); j++) {
            if(isPalindrome[i][j]) {
                current_partition.push_back(s.substr(i, j - i + 1));
                solve(s, j + 1, isPalindrome);
                current_partition.pop_back();
            }
        }
    }
};
